File: EDUCATIVE/INVERVAL_PROBLEMS/merge_intervals.java


import java.util.*;

class Interval {

    int start;
    int end;

    public Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
};


class MergeIntervals {

    public static List<Interval> mergeInterval (List<Integer> intervals) {
        if (intervals.size() < 2) {
            return intervals;
        }

        Collections.sort(intervals, (a,b) -> Integer.compare(a.start, b.start));
        List<Internal> mergedIntervals = new LinkedList<>();

        Iterator<Interval> internalPtr = intervals.iterator();
        Internal interval = internalPtr.next();

        int start = interval.start;
        int end  = interval.end;

        while(internalPtr.hasNext()) {
            interval = internalPtr.next();
            if(interval.start <= end) {
                // overlapping interval , adjust the end
                end = Math.max(interval.end, end);
            } else {
                // non overlapping internal
                mergedIntervals.add(new Interval(start, end));
                start = interval.start;
                end = interval.end;
            }
        }

        // add the last interval
        mergedIntervals.add(new Interval(start, end));

    }
}

File: EDUCATIVE/INVERVAL_PROBLEMS/merge_intervals.py
from __future__ import print_function

class Interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end

    def print_interval(self):
        print ("[" + str(self.start) + " , " + str(self.end) + "]", end='')

def merge(intervals):
  if len(intervals) < 2:
      return intervals

  #sort intervals by the start time
  intervals.sort(key=lambda x : x.start)


  mergedIntervals = []
  start = intervals[0].start
  end = intervals[0].end

  for i in range(1, len(intervals)):
      interval = intervals[i]
      if interval.start <= end: # overlapping interval, adjust the end
          end = max(interval.end, end)
      else: # non overlapping interval, add the previous interval
          mergedIntervals.append(Interval(start, end))
          start = interval.start
          end   = interval.end
  # add last interval
  mergedIntervals.append(Interval(start, end))
  return mergedIntervals


def main():
  print("Merged intervals: ", end='')
  for i in merge([Interval(1, 4), Interval(2, 5), Interval(7, 9)]):
    i.print_interval()
  print()

  print("Merged intervals: ", end='')
  for i in merge([Interval(6, 7), Interval(2, 4), Interval(5, 9)]):
    i.print_interval()
  print()

  print("Merged intervals: ", end='')
  for i in merge([Interval(1, 4), Interval(2, 6), Interval(3, 5)]):
    i.print_interval()
  print()


main()